### PR TITLE
feat: multi-core dynamic-batch matmul

### PR DESCRIPTION
**Pull request by fiskrt**
## Dynamic-batch matmul with fixed shapes
**`M,K,N` is known at compile-time while `bs` is not.**

For now we assume `A: [bs, M, K]`, `B: [K, N]`, `C: [bs, M, N] `and we do batched matmul (`B` is broadcasted).

The reduction over the k dim is done in chunks of size `BASEK`. (So for `A` tiles are `[M, BASEK]` and for `B: [BASEK, N]`)

Each core c=0,1,...,C-1 get allocated `ceil(bs, C)` batches (this is bad load-balancing for e.g. bs=9 C=8 since some cores does 0 core 0-3 does 2 batches, core 4 does 1 batch, and cores 5-7 does 0 batches each.)

TODO: every multiplcation will use the same B, so make sure we keep it in fast mem

### CODE REVIEW

File: examples/aot/matmul_dynbatch_multicore/run_matmul.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    bs = 5
+    m, k, n = 128, 128, 128
+    torch.manual_seed(0)
+    a = torch.rand((bs, m,k), device=device, dtype=dtype)
+    b = torch.rand((k,n), device=device, dtype=dtype)
+    c = torch.zeros((bs, m, n), device=device, dtype=dtype)
+
+    matmul_func = load_lib("./matmul_kernel.so")
+    matmul_func(c, a, b, batch_size=a.shape[0], block_dim=2)
```
[the remark this answers is not in the document]
Comment (fiskrt) [reply examples/aot/matmul_dynbatch_multicore/run_matmul.py]: image
